File: heidi-credentials/src/commonMain/kotlin/ch/ubique/heidi/credentials/W3C.kt
/* Copyright 2025 Ubique Innovation AG

Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
 */

package ch.ubique.heidi.credentials

import ch.ubique.heidi.credentials.sdjwt.SdJwtVcSignatureResolver
import ch.ubique.heidi.util.extensions.asObject
import ch.ubique.heidi.util.extensions.asString
import ch.ubique.heidi.util.extensions.get
import ch.ubique.heidi.util.extensions.json
import kotlinx.serialization.Serializable
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.json.Json
import uniffi.heidi_credentials_rust.*
import uniffi.heidi_crypto_rust.base64UrlEncode
import uniffi.heidi_util_rust.Value

sealed class W3C {
    companion object {
        val W3C_FORMATS: Array<String> = arrayOf("vc+sd-jwt")

        // NOTE: In the future other W3C formats can be supported
        fun parse(credential: String): W3C =
            SdJwt(parseW3cSdJwt(credential))
    }

    abstract val docType: String

    abstract fun asJson(): Value

    abstract fun presentation(): SdJwtBuilder

    abstract fun getOriginalNumClaims(): Int

    abstract fun getNumDisclosed(): Int

    abstract fun isSignatureValid(): Boolean

    data class SdJwt(val inner: W3cSdJwt) : W3C() {
        override val docType: String = inner.doctype

        override fun asJson(): Value =
            inner.json

        override fun presentation(): SdJwtBuilder =
            SdJwtBuilder.fromW3c(inner)

        override fun getOriginalNumClaims(): Int =
            inner.numDisclosures.toInt()

        override fun getNumDisclosed(): Int =
            inner.numDisclosures.toInt()

        override fun isSignatureValid(): Boolean
            = SdJwtVcSignatureResolver.isSignatureValid(inner.originalJwt)

        companion object {
            fun create(
                claims: Value,
                disclosures: List<ClaimsPointer>,
                keyId: String,
                key: SignatureCreator,
                pubKeyJwk: Value?
            ): SdJwt? {
                if (claims !is Value.Object) {
                    return null;
                }

                val header = Header(alg = key.alg(), kid = keyId)
                val keyClaims = claims.asObject()!!.toMutableMap()

                if (pubKeyJwk != null) {
                    keyClaims.put("cnf", Value.Object(mapOf("jwk" to pubKeyJwk)))
                }

                val claimObject = Value.Object(keyClaims)

                val sdJwt = createDisclosureForObject(claimObject, disclosures, 1)

                val headerEncoded = base64UrlEncode(
                    Json.encodeToString(Header.serializer(), header).encodeToByteArray()
                )
                val sdJwtDisclosure = sdJwt.getOrNull() ?: return null
                val bodyEncoded = base64UrlEncode(
                    Json.encodeToString(sdJwtDisclosure.disclosedObject).encodeToByteArray()
                )
                val msgPayload = "$headerEncoded.$bodyEncoded"
                val signature = base64UrlEncode(key.sign(msgPayload.encodeToByteArray()))
                val jwt = "$msgPayload.$signature"
                val disclosureString = sdJwtDisclosure.disclosure.joinToString("~")

                return SdJwt(parseW3cSdJwt("$jwt~$disclosureString~"))
            }
        }
    }

    data class OpenBadge303(private val data: Data) : W3C() {
        /**
         * Helper class to store precomputed information such that we don't have to recompute it
         * again every time we load the credential.
         */
        @Serializable
        data class Data(
            val credential: OpenBadges303Credential,

            // NOTE: To check the credential validity network requests must be potentially made, as
            //       such it is favorable to cache this value. But do note, that if the credential
            //       expires, this value won't be updated.
            val isValid: Boolean,
        )

        override val docType: String
            get() = data.credential.data.types.firstOrNull { it != "VerifiableCredential" }
                ?: "VerifiableCredential"

        val types: List<String>
            get() = data.credential.data.types

        val name: String?
            get() = when (val name = data.credential.data.name) {
                is LocalizableString.ManyLvo -> name.v1.firstOrNull()?.value
                is LocalizableString.OneLvo -> name.v1.value
                is LocalizableString.String -> name.v1
                null -> null
            }

        val achievement: Value
            get() = data.credential.data.credentialSubject?.get("achievement") ?: Value.Null

        val issuerName: String?
            get() = data.credential.data.issuer?.get("name")?.asString()

        val pngBytes: ByteArray = data.credential.imageBytes

        private val dataAsJson: Value by lazy {
            w3cCredentialAsJson(data.credential.originalData)
        }

        val originalString: String
            get() = this.data.credential.original

        override fun asJson(): Value = dataAsJson

        override fun presentation(): SdJwtBuilder {
            throw Exception("Use `asVerifiablePresentation()` instead")
        }

        fun asVerifiablePresentation(): Result<String> = runCatching {
            val proof = Value.Object(mapOf(
                "@context" to Value.Array(listOf(
                    Value.String("https://www.w3.org/2018/credentials/v1"),
                    Value.String("https://w3id.org/security/data-integrity/v2")
                )),
                "type" to Value.Array(listOf(
                    Value.String("VerifiablePresentation")
                )),
                "verifiableCredential" to Value.Array(listOf(dataAsJson)),
            ))

            json.encodeToString(proof)
        }

        override fun getOriginalNumClaims(): Int = 0
        override fun getNumDisclosed(): Int = 0

        override fun isSignatureValid(): Boolean = data.isValid

        fun serialized(): String =
            Json.encodeToString(this.data)

        fun asW3CCredential(): W3cVerifiableCredential =
            data.credential.originalData

        companion object {
            val OPEN_BADGE_FORMATS: Array<String> = arrayOf("ldp_vc")

            suspend fun parse(bytes: ByteArray): OpenBadge303 {
                val credential = parseOpenBadges303Credential(bytes, listOf(
                    "https://purl.imsglobal.org/spec/ob/v3p0/context-3.0.3.json",
                    "https://purl.imsglobal.org/spec/ob/v3p0/extensions.json"
                ))

                val isValid = verifySecuredDocumentString(credential.original)

                return OpenBadge303(Data(credential, isValid))
            }

            fun parseSerialized(credential: String): OpenBadge303 =
                OpenBadge303(Json.decodeFromString(credential))
        }
    }
}

File: heidi-credentials/rust/src/w3c.rs
/* Copyright 2025 Ubique Innovation AG

Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
 */

use std::{
    collections::HashMap,
    fmt::{Display, Formatter},
    sync::Arc,
};

use heidi_util_rust::value::Value;
use serde::Serialize;
use serde_json::Value as JsonValue;

use crate::{
    claims_pointer::Selector,
    sdjwt_util::{self, Disclosure, DisclosureTree, SdJwtDecodeError},
};

#[derive(Debug, Clone, uniffi::Error)]
pub enum W3CParseError {
    SdJwtError(SdJwtDecodeError),
    NoType,
}

impl Display for W3CParseError {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        f.write_fmt(format_args!("{:?}", self))
    }
}

#[derive(Debug, Clone, uniffi::Record, Serialize)]
pub struct W3CSdJwt {
    /// The credential document type
    pub doctype: String,

    /// The credential as a JSON object
    pub json: Value,

    /// The original JWT (SD-JWT without the disclosures and kb-jwt)
    pub original_jwt: String,

    /// The original SD-JWT
    pub original_sdjwt: String,

    /// The disclosure map, mapping sd-hashes to decoded disclosures
    pub disclosure_map: HashMap<String, Disclosure>,

    /// The disclosure tree, mapping paths to disclosures
    pub disclosure_tree: DisclosureTree,

    /// The number of disclosures in the disclosure map
    pub num_disclosures: u32,
}

impl W3CSdJwt {
    pub fn get(&self, selector: Arc<dyn Selector>) -> Option<Vec<Value>> {
        selector.select(self.json.clone()).ok()
    }
}

#[uniffi::export]
pub fn parse_w3c_sd_jwt(credential: &str) -> Result<W3CSdJwt, W3CParseError> {
    let decoded = sdjwt_util::decode_sdjwt(credential).map_err(W3CParseError::SdJwtError)?;

    let types = match decoded.claims.get("type") {
        Some(JsonValue::Array(types)) => types
            .iter()
            .filter_map(|t| t.as_str().map(|s| s.to_string()))
            .collect::<Vec<_>>(),
        Some(JsonValue::String(r#type)) => vec![r#type.clone()],
        _ => return Err(W3CParseError::NoType),
    };

    let Some(doctype) = types
        .iter()
        .find(|t| t.as_str() != "VerifiableCredential")
        .or(types.first())
    else {
        return Err(W3CParseError::NoType);
    };

    Ok(W3CSdJwt {
        doctype: doctype.clone(),
        json: decoded.claims.into(),
        original_sdjwt: decoded.original_sdjwt,
        original_jwt: decoded.original_jwt,
        disclosure_map: decoded.disclosure_map,
        disclosure_tree: decoded.disclosure_tree,
        num_disclosures: decoded.num_disclosures as u32,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_w3c_sd_jwt() {
        let credential = "eyJraWQiOiJFeEhrQk1XOWZtYmt2VjI2Nm1ScHVQMnNVWV9OX0VXSU4xbGFwVXpPOHJvIiwiYWxnIjoiRVMyNTYifQ.eyJpYXQiOjE3NDU3NzY3MTMsImV4cCI6MTc0Njk4NjMxMywiX3NkX2FsZyI6InNoYS0yNTYiLCJAY29udGV4dCI6WyJodHRwczovL3d3dy53My5vcmcvbnMvY3JlZGVudGlhbHMvdjIiLCJodHRwczovL3d3dy53My5vcmcvbnMvY3JlZGVudGlhbHMvZXhhbXBsZXMvdjIiXSwiaXNzdWVyIjoiaHR0cHM6Ly91bml2ZXJzaXR5LmV4YW1wbGUvaXNzdWVycy81NjUwNDkiLCJ2YWxpZEZyb20iOiIyMDEwLTAxLTAxVDAwOjAwOjAwWiIsImNyZWRlbnRpYWxTdWJqZWN0Ijp7ImRlZ3JlZSI6eyJuYW1lIjoiQmFjaGVsb3Igb2YgU2NpZW5jZSBhbmQgQXJ0cyIsIl9zZCI6WyJEUkg1aWVsZHdHNXJPMlVQNXlYYlBXWHNTaFFNSmxESlJfZlFVbmhZVDNFIl19LCJfc2QiOlsiUzRvTGpDb0dNckpuMnFFR2lXY1JNNmdFNGZ6cVVFcVIzNC1FOWdjZzIyWSJdfSwiX3NkIjpbIlZtWnFMMkpKUFB0RDk2TmxwNE43TzFRMXhFRmNMZ1hCVzVfQWFGQXp4Sm8iLCJaYTdxRkpZSnRSTExSOFNRT1VUYUxwaDZBY21QSGlYVkc5Ni03Wnp3MEtJIl19.ypl46Q1EqUERV-IUUS_-qGoAESfv_WdXwtHOk2vX7QTZNFf0NNfg-w2OR8JPRe97kZBDQLuBZKPJhBXdFjbSwg~WyIxeDVielRkZXhsLW4zWVVIQXF5ZUxBIiwgImlkIiwgImh0dHA6Ly91bml2ZXJzaXR5LmV4YW1wbGUvY3JlZGVudGlhbHMvMzczMiJd~WyJablVReVZXRmo0UlFfTHFmOVBkbmN3IiwgInR5cGUiLCBbIlZlcmlmaWFibGVDcmVkZW50aWFsIiwgIkV4YW1wbGVEZWdyZWVDcmVkZW50aWFsIl1d~WyI5TG1nOHhaUVJxWEZZaVRlV0hRZjV3IiwgImlkIiwgImRpZDpleGFtcGxlOmViZmViMWY3MTJlYmM2ZjFjMjc2ZTEyZWMyMSJd~WyJZMVBDaVA3YnJ3TjFHMEVMWmJXRlZRIiwgInR5cGUiLCAiRXhhbXBsZUJhY2hlbG9yRGVncmVlIl0~";
        let parsed = parse_w3c_sd_jwt(credential).unwrap();

        assert_eq!(parsed.doctype, "ExampleDegreeCredential");
    }
}
